src/providers/VoiceFocusProvider/index.tsx

```diff
@@ -0,0 +1,167 @@
+
+interface Props {
+  /** Determines how you want the Amazon Voice Focus to behave. This spec is used to derive a runtime configuration when a transformer is created. */
+  spec?: VoiceFocusSpec;
+  /**
+  * A set of options that can be supplied when creating an Amazon Voice Focus device.
+  * For more info, you can go to https://aws.github.io/amazon-chime-sdk-js/interfaces/voicefocusdeviceoptions.html
+  */
+  options?: VoiceFocusDeviceOptions;
+}
+
+interface VoiceFocusState {
+  isVoiceFocusSupported: boolean | undefined;
+  addVoiceFocus: (device: Device) => Promise<Device | VoiceFocusTransformDevice>;
+}
+
+const VoiceFocusContext = createContext<VoiceFocusState | null>(null);
+
+const VoiceFocusProvider: React.FC<Props> = ({
+  spec,
+  options,
+  children
+}) => {
+  const [isVoiceFocusSupported, setIsVoiceFocusSupported] = useState<boolean | undefined>(undefined);
+  const [voiceFocusDevice, setVoiceFocusDevice] = useState<VoiceFocusTransformDevice | null>(null);
+  const [voiceFocusTransformer, setVoiceFocusTransformer] = useState<VoiceFocusDeviceTransformer | null>(null);
+
+  // Check whether the value of spec is undefined or empty object
+  const vfSpec = useMemoCompare(spec, (prev: any, next: any) => {
+    const FIXED_EMPTY = {};
+    next === undefined ? FIXED_EMPTY : Object.keys(next).length === 0 ? FIXED_EMPTY : next;
+    return prev && Object.keys(prev).length === 0 && Object.keys(next).length === 0;
+  });
+
+  function useMemoCompare(next: any, compare: any) {
+    const previousRef = useRef();
+    const previous = previousRef.current;
+    const isEqual = compare(previous, next);
+
+    useEffect(() => {
+      if (!isEqual) {
+        previousRef.current = next;
+      }
+    });
+
+    return isEqual ? previous : next;
+  }
+
+  const addVoiceFocus = useCallback(async (device: Device): Promise<Device | VoiceFocusTransformDevice> => {
+    if (!voiceFocusDevice) {
+      return createVoiceFocusDevice(device);
+    }
+    const vf = await voiceFocusDevice.chooseNewInnerDevice(device);
+    setVoiceFocusDevice(vf);
+    return vf;
+  }, [isVoiceFocusSupported]);
+
+  async function createVoiceFocusDevice(inner: Device): Promise<Device | VoiceFocusTransformDevice> {
+    if (!isVoiceFocusSupported) {
+      return inner;
+    }
+    try {
+      const transformer = await getVoiceFocusDeviceTransformer();
+      const device = await transformer?.createTransformDevice(inner);
+      if (device) {
+        setVoiceFocusDevice(device);
+        return device;
+      }
+    } catch (e) {
+      console.warn('Amazon Voice Focus is not supported.', e);
+    }
+    return inner;
+  }
+
+  async function getVoiceFocusDeviceTransformer(): Promise<VoiceFocusDeviceTransformer> {
+    if (voiceFocusTransformer) {
+      return voiceFocusTransformer;
+    }
+
+    const transformer = await VoiceFocusDeviceTransformer.create(spec, options);
+    setVoiceFocusTransformer(transformer);
+    return transformer;
```

Comment:
Can't this race with the init? (Not the first time, because the support boolean won't have been set, but subsequently…)